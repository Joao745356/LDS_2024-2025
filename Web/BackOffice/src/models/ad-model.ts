/**
 * Interface representing an Advertisement.
 */
export interface Ad {
    /**
     * Unique identifier for the advertisement.
     */
    id: number;
  
    /**
     * ID of the admin who created or owns the advertisement.
     */
    adminID: number;
  
    /**
     * Indicates whether the advertisement is currently active.
     */
    isActive: boolean;
  
    /**
     * The date when the advertisement starts.
     * This field defines the start date of the ad's validity.
     */
    startDate: Date;
  
    /**
     * The date when the advertisement ends.
     * This field defines the end date of the ad's validity.
     */
    endDate: Date;
  
    /**
     * URL or path to the advertisement's image file.
     * This could be used for displaying the ad's visual content.
     */
    adFile: string;
  }
  