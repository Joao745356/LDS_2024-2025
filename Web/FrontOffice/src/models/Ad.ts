export interface Ad{
    id: number,
    adminId: number,
    isActive: boolean,
    startDate: Date,
    endDate: Date,
    adFile: string
}