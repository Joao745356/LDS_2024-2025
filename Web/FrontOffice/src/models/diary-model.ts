
export interface Diary {
    id: number,
    UserPlantId: number,
    title: string,
    Datetime: Date,
}