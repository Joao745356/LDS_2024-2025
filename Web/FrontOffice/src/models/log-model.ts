

export interface Log{
    id:number,
    diaryId : number,
    logDate : Date,
    logDescription : string,
}