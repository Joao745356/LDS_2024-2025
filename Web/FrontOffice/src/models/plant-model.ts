import { LightLevel } from "./enums/LightLevel";
import { WaterLevels } from "./enums/WaterLevels";
import { PlantType } from "./enums/PlantType";
import { ExperienceLevels } from "./enums/ExperienceLevels";

export interface Plant {
  id: number;
  name: string;
  type: PlantType ;
  expSuggested: ExperienceLevels;
  waterNeeds: WaterLevels;
  luminosityNeeded: LightLevel;
  description: string;
  plantImage: string;
}