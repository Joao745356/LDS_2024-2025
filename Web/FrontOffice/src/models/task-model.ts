export interface Task{
    id:number,
    adminId:number
    plantId : number,
    taskName : string,
    taskDescription : string,
}