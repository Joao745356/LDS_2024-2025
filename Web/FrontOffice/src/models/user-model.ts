export interface User {
    id: number,
    username: string,
    email: string,
    contact: string,
    rolePaid: boolean,
    location: string,
    careExperience: "Beginner" | "Intermediate" | "Expert",
    waterAvailability: "Low" | "Medium" | "High",
    luminosityAvailability: "Low" | "Medium" | "High",
    userAvatar: string

}

export interface UserPreferences {
    waterAvailability: "Low" | "Medium" | "High",
    luminosityAvailability: "Low" | "Medium" | "High",
    careExperience: "Beginner" | "Intermediate" | "Expert",
}