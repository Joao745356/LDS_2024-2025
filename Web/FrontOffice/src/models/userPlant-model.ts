import { Plant } from "./plant-model";

export interface UserPlant {
  id?: number;
  userId?: number;
  plantId: number;
  plant : Plant,
}
