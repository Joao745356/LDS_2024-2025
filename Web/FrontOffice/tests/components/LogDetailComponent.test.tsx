import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi, describe, it, expect } from "vitest";
import '@testing-library/jest-dom';
import LogDetailPopup from "../../src/components/Diary/LogDetailComponent";
import { Log } from "../../src/models/log-model";
import React from "react";

describe("LogDetailPopup", () => {
  // Mock log data
  const mockLog: Log = {
    id: 1,
    logDate: new Date('2023-05-15'),
    logDescription: "This is a detailed log entry about my plant's growth.",
    diaryId: 1
  };

  // 1. Rendering Tests


  it("does not render when log is not provided", () => {
    const { container } = render(
      <LogDetailPopup 
        onClose={() => {}} 
        log={undefined as any} 
      />
    );

    expect(container.firstChild).toBeNull();
  });

  it("renders log details when open", () => {
    render(
      <LogDetailPopup 
        onClose={() => {}} 
        log={mockLog} 
      />
    );

    // Check log date
    expect(screen.getByText("Mon May 15 2023")).toBeInTheDocument();

    // Check log description
    expect(screen.getByText("This is a detailed log entry about my plant's growth.")).toBeInTheDocument();

    // Check close button exists
    expect(screen.getByText("Close")).toBeInTheDocument();
  });

  // 2. Interaction Tests
  it("calls onClose when close button is clicked", async () => {
    const mockOnClose = vi.fn();

    render(
      <LogDetailPopup 
        onClose={mockOnClose} 
        log={mockLog} 
      />
    );

    const closeButton = screen.getByText("Close");
    await userEvent.click(closeButton);

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  // 3. Accessibility and UI Tests
  it("has correct aria attributes", () => {
    render(
      <LogDetailPopup 
        onClose={() => {}} 
        log={mockLog} 
      />
    );

    const popup = screen.getByText(mockLog.logDescription).closest('div');
    expect(popup).toHaveClass("bg-white");
    expect(popup).toHaveClass("rounded-lg");
  });

  // 4. Date Formatting Test
  it("formats date correctly", () => {
    render(
      <LogDetailPopup 
        onClose={() => {}} 
        log={mockLog} 
      />
    );

    // Check if date is formatted as expected
    expect(screen.getByText("Mon May 15 2023")).toBeInTheDocument();
  });

  // 5. Snapshot Test
  it("matches snapshot", () => {
    const { asFragment } = render(
      <LogDetailPopup 
        onClose={() => {}} 
        log={mockLog} 
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});