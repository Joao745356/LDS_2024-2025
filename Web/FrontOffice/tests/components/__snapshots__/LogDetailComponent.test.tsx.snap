// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LogDetailPopup > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 flex justify-center items-center bg-black bg-opacity-50 z-50"
  >
    <div
      class="bg-white p-6 rounded-lg shadow-lg max-w-4xl w-full"
    >
      <p
        class="text-gray-700 mb-4"
      >
        Mon May 15 2023
      </p>
      <p
        class="text-lg"
      >
        This is a detailed log entry about my plant's growth.
      </p>
      <div
        class="mt-4 flex justify-end"
      >
        <button
          class="px-4 py-2 bg-red-500 text-white rounded-lg hover:bg-red-600"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
